### PR TITLE
fix #21243: Allow selection of multiple parts in Instruments dialog

### LINKED ISSUE #21243: Dynamics overriding accents on the Muse Sounds snare drum in Musescore 4.2
### Issue type

Muse Sounds bug

### Bug description

Dynamics seem to be overriding the snare drum accents with the muse sounds snare drum. When I try to make a snare drum part with labeled dynamics, mp for example, none of the accents will be played. When I try and play the line after removing the dynamic marking, it plays the accents. This is currently preventing me from writing snare drum parts with changing dynamics and interesting rhythms.

### Steps to reproduce

1. Create a new score with a snare drum part
2. Put in a few notes, some with accents and some without
3. Add a dynamic marking
4. Playback the part

### Screenshots/Screen recordings

_No response_

### MuseScore Version

4.2

### Regression

Yes, this used to work in a previous version of MuseScore 4.x

### Operating system

Windows 11

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by allardnaber**
Fix to allow selection of multiple parts/staves in the Instruments to move or remove multiple parts or staves at once. Apart from the possibility to select multiple parts it will automatically select the parts/staves that are involved (selecting children from multiple parts will select all parts and staves, selecting a part will select all staves, as moving and removing already applies to all children).
Methods for moving a part/staff could be optimized but I have touched them only where necessary.

Open for all feedback, as this is the first contribution. Commits will probably need to be squashed in a single commit?

### CODE REVIEW

**Comment by jthistle** [comment]
Welcome! Congratulations on your first MuseScore pull request. For PRs, we tend to prefer if all the commits are squashed into a single commit with the issue number in the commit message. For example, your PR title would do: `fix #21243: Allow selection of multiple parts in Instruments dialog`.
To do this, you can do an interactive rebase with `git rebase HEAD~6 -i` on your branch. Change the first commit from `pick` to `reword` and change the next 5 from `pick` to `squash`. Then, just carefully read and follow the instructions that git provides to you to change the commit message.

([Node for reference](https://musescore.org/en/node/21243))

File: mscore/instrwidget.cpp [hunk 1]
```diff
@@ -478,6 +480,139 @@ void InstrumentsWidget::genPartList(Score* cs)
...
+                                    int plusIdx = delta;
+                                    QTreeWidgetItem* nextParent = partiturList->topLevelItem(idx + plusIdx);
+                                    while (nextParent && nextParent->isHidden()) {
+                                          plusIdx += delta;
+                                          nextParent = partiturList->topLevelItem(idx + plusIdx);
+                                    }
+                                    partiturList->insertTopLevelItem(idx + plusIdx, item1);
+                                    // after-re-insertion, recreate each combo and set its index
+                                    for (itemIdx=0; itemIdx < numOfStaffListItems; ++itemIdx) {
+                                          StaffListItem* staffItem = static_cast<StaffListItem*>(item1->child(itemIdx));
```
Comment (jthistle) [inline mscore/instrwidget.cpp]: `static_cast` should be avoided if possible. Does `toStaffListItem(item1->child(itemIdx)` work instead? If not, you can stick with static_cast.

File: mscore/instrwidget.cpp [hunk 2]
```diff
@@ -478,6 +480,139 @@ void InstrumentsWidget::genPartList(Score* cs)
...
+                  int idx = delta > 0 ? parent->childCount() - 1 : 0;
+                  for ( ; (item = parent->child(idx)); idx += -delta) {
+                        if (!itemsToMove.contains(item))
+                              continue;
+                  
+                        int idx = parent->indexOfChild(item);
+                        int n = parent->childCount();
+                        // if staff item is not last of its part, move one slot down in part
+                        if ((idx + delta < n) && (idx + delta >= 0)) {
+                              StaffListItem* item1 = static_cast<StaffListItem*>(parent->takeChild(idx));
```
Comment (jthistle) [inline mscore/instrwidget.cpp]: Same as above.

File: mscore/instrwidget.cpp [hunk 3]
```diff
@@ -590,85 +723,93 @@ void InstrumentsWidget::on_addButton_clicked()
...
+                        }
                   }
-            int idx = partiturList->indexOfTopLevelItem(item);
-            if (((PartListItem*)item)->op == ListItemOp::ADD) {
-                  partiturList->blockSignals(true);
-                  delete item; // fires selectionChanged too early (item is still in view)
-                  partiturList->blockSignals(false);
-                  emit on_partiturList_itemSelectionChanged(); // fire manually (item gone by now)
+                  if (countTopLevelItems == 1) {
+                        //Q_ASSERT(false); // shouldn't get here as (remove button disabled when one item left)
```
Comment (jthistle) [inline mscore/instrwidget.cpp]: This comment can be removed, I think.

File: mscore/instrwidget.cpp [hunk 4]
```diff
@@ -478,6 +480,139 @@ void InstrumentsWidget::genPartList(Score* cs)
...
+                                    int itemIdx;
+                                    for (itemIdx=0; itemIdx < numOfStaffListItems; ++itemIdx)
+                                          staffIdx[itemIdx] = (static_cast<StaffListItem*>(item1->child(itemIdx)))->staffTypeIdx();
+                                    // do not consider hidden ones
+                                    int plusIdx = delta;
+                                    QTreeWidgetItem* nextParent = partiturList->topLevelItem(idx + plusIdx);
+                                    while (nextParent && nextParent->isHidden()) {
+                                          plusIdx += delta;
+                                          nextParent = partiturList->topLevelItem(idx + plusIdx);
+                                    }
```
Comment (jthistle) [inline mscore/instrwidget.cpp]: Indentation: closing brackets are indented one tab further than the start of the block, e.g. this should have an extra tab (6 spaces) before it.

File: mscore/instrwidget.cpp [hunk 5]
```diff
@@ -590,85 +723,93 @@ void InstrumentsWidget::on_addButton_clicked()
...
+                              partiturList->takeTopLevelItem(partiturList->indexOfTopLevelItem(parent));
+                              delete parent;
                         }
-                  else {
-                        parent->takeChild(parent->indexOfChild(item));
-                        delete item;
+                        else {
+                              parent->takeChild(parent->indexOfChild(item));
+                              delete item;
                         }
```
Comment (jthistle) [inline mscore/instrwidget.cpp]: Indentation.

File: mscore/instrwidget.cpp [hunk 6]
```diff
@@ -590,85 +723,93 @@ void InstrumentsWidget::on_addButton_clicked()
...
+                              StaffListItem* sli2 = static_cast<StaffListItem*>(parent->child(idx+2));
+                              if (sli->linked() && !(sli2 && sli2->linked())) {
+                                    sli->setLinked(false);
+                                    partiturList->update();
                               }
                         }
                   }
-            static_cast<PartListItem*>(parent)->updateClefs();
-            partiturList->setCurrentItem(parent);
+                  static_cast<PartListItem*>(parent)->updateClefs();
```
Comment (jthistle) [inline mscore/instrwidget.cpp]: Another unneeded static cast? I haven't actually checked if we have a `toPartListItem` function , so if it doesn't exist, ignore all of this :)

File: mscore/instrwidget.cpp [hunk 7]
```diff
@@ -590,85 +723,93 @@ void InstrumentsWidget::on_addButton_clicked()
...
-                  removeButton->setEnabled(false); // nevertheless, handle gracefully in release builds
-                  emit completeChanged(false);
-                  return;
+            else { // PARENTS_SELECTED
+                  int countTopLevelItems = 0;
+                  QTreeWidgetItem* it;
+                  for (int idx = 0; (it = partiturList->topLevelItem(idx)); ++idx) {
+                        if (!it->isHidden()) {
+                              countTopLevelItems++;
+                        }
```
Comment (jthistle) [inline mscore/instrwidget.cpp]: Indentation.

File: mscore/instrwidget.cpp [hunk 8]
```diff
@@ -590,85 +723,93 @@ void InstrumentsWidget::on_addButton_clicked()
...
                   }
```
Comment (jthistle) [inline mscore/instrwidget.cpp]: Indentation.

File: mscore/instrwidget.cpp [hunk 9]
```diff
@@ -590,85 +723,93 @@ void InstrumentsWidget::on_addButton_clicked()
...
-                  partiturList->blockSignals(true);
-                  delete item; // fires selectionChanged too early (item is still in view)
-                  partiturList->blockSignals(false);
-                  emit on_partiturList_itemSelectionChanged(); // fire manually (item gone by now)
+                  if (countTopLevelItems == 1) {
+                        //Q_ASSERT(false); // shouldn't get here as (remove button disabled when one item left)
+                        removeButton->setEnabled(false); // nevertheless, handle gracefully in release builds
+                        emit completeChanged(false);
+                        return;
                   }
```
Comment (jthistle) [inline mscore/instrwidget.cpp]: Indentation...

File: mscore/instrwidget.cpp [hunk 10]
```diff
@@ -590,85 +723,93 @@ void InstrumentsWidget::on_addButton_clicked()
...
-            // select an item, do not consider hidden ones
-            int plusIdx = 0;
-            QTreeWidgetItem* nextParent = partiturList->topLevelItem(idx + plusIdx);
-            while (nextParent && nextParent->isHidden()) {
-                  plusIdx++;
-                  nextParent = partiturList->topLevelItem(idx + plusIdx);
+                  else {
+                        ((PartListItem*)item)->op = ListItemOp::I_DELETE;
+                        item->setHidden(true);
                   }
```
Comment (jthistle) [inline mscore/instrwidget.cpp]: Indentation...

File: mscore/instrwidget.cpp [hunk 11]
```diff
@@ -590,85 +723,93 @@ void InstrumentsWidget::on_addButton_clicked()
...
-            if(!nextParent) { // couldn't find one after, check before
-                  plusIdx = 1;
-                  nextParent = partiturList->topLevelItem(idx - plusIdx);
+                  // select an item, do not consider hidden ones @TODO: remove this and let QTreeWidget handle it.
+                  int plusIdx = 0;
+                  QTreeWidgetItem* nextParent = partiturList->topLevelItem(idx + plusIdx);
                   while (nextParent && nextParent->isHidden()) {
                         plusIdx++;
+                        nextParent = partiturList->topLevelItem(idx + plusIdx);
+                  }
```
Comment (jthistle) [inline mscore/instrwidget.cpp]: Indentation.

File: mscore/instrwidget.cpp [hunk 12]
```diff
@@ -590,85 +723,93 @@ void InstrumentsWidget::on_addButton_clicked()
...
+                  if(!nextParent) { // couldn't find one after, check before
+                        plusIdx = 1;
                         nextParent = partiturList->topLevelItem(idx - plusIdx);
+                        while (nextParent && nextParent->isHidden()) {
+                              plusIdx++;
+                              nextParent = partiturList->topLevelItem(idx - plusIdx);
                         }
```
Comment (jthistle) [inline mscore/instrwidget.cpp]: Indentation.

File: mscore/instrwidget.cpp [hunk 13]
```diff
@@ -590,85 +723,93 @@ void InstrumentsWidget::on_addButton_clicked()
...
                   }
-            partiturList->setCurrentItem(nextParent);
+                  }
```
Comment (jthistle) [inline mscore/instrwidget.cpp]: Two braces shouldn't have to appear above each other like this :)

... it's indentation again :)

File: mscore/instrwidget.h [hunk 14]
```diff
@@ -31,6 +31,7 @@ class InstrumentGenre;
 
 enum class ListItemOp : char { KEEP, I_DELETE, ADD, UPDATE };
 enum { PART_LIST_ITEM = QTreeWidgetItem::UserType, STAFF_LIST_ITEM };
+enum PartiturSelectionType { PARENTS_SELECTED, CHILDREN_SELECTED, NONE_SELECTED };
```
Comment (jthistle) [inline mscore/instrwidget.h]: `enum class : char` would be preffered, I think? @dmitrio95 could confirm, possibly.

File: mscore/revision.h [hunk 15]
```diff
@@ -1 +1 @@
-3543170
+a504e39
```
Comment (jthistle) [inline mscore/revision.h]: I don't know how this got included, .gitignore should ignore revision.h. Nevertheless, this file should not be in the commited files.

File: mscore/instrwidget.cpp (hunk 1 above)
Comment (allardnaber) [reply mscore/instrwidget.cpp]: This was in the original code, only moved to a separate method. I'll have a look.

File: mscore/instrwidget.cpp (hunk 2 above)
Comment (allardnaber) [reply mscore/instrwidget.cpp]: Same here

**Comment by jthistle** [comment]
Ah... you seem to have got the commits into a bit of a pickle. Can you do:

`git reset 746c4f3e59edb1999b3400a8d3dd31cac2bdff50 --hard`

on your development branch, please? Then, follow my instructions again, as you seem to have already done. Instead of doing a normal push, do `git push origin fix-21243-multiple-selection -f`. The `-f` is important, it allows you to force-push the branch with commits in a different order with different names.

File: mscore/instrwidget.cpp (hunk 3 above)
Comment (allardnaber) [reply mscore/instrwidget.cpp]: You're right, I missed this one

File: mscore/instrwidget.cpp (hunk 4 above)
Comment (allardnaber) [reply mscore/instrwidget.cpp]: Thanks, it's very hard to get this to work properly in xcode...

File: mscore/instrwidget.cpp [hunk 16]
```diff
@@ -478,6 +480,139 @@ void InstrumentsWidget::genPartList(Score* cs)
...
+            partiturList->selectionModel()->clear();
+            
+            QTreeWidgetItem* item;
+            if (selectionType == PARENTS_SELECTED) {
+                  int idx = delta > 0 ? partiturList->topLevelItemCount() - 1 : 0;
+                  for ( ; (item = partiturList->topLevelItem(idx)); idx += -delta) {
+                        if (!itemsToMove.contains(item))
+                              continue;
+                        
+                              bool isExpanded = partiturList->isItemExpanded(item);
```
Comment (JoshuaBonn1) [inline mscore/instrwidget.cpp]: Seems the indentation issue starts here. This spot and on should match the indentation of the above if statement (subtract 6 spaces).

File: mscore/instrwidget.cpp [hunk 17]
```diff
@@ -478,6 +480,139 @@ void InstrumentsWidget::genPartList(Score* cs)
...
+                                    int numOfStaffListItems = item1->childCount();
+#if (!defined (_MSCVER) && !defined (_MSC_VER))
+                                    int staffIdx[numOfStaffListItems];
+#else
+                                    // MSVC does not support VLA. Replace with std::vector. If profiling determines that the
+                                    //    heap allocation is slow, an optimization might be used.
+                                    std::vector<int> staffIdx(numOfStaffListItems);
+#endif
+                                    int itemIdx;
+                                    for (itemIdx=0; itemIdx < numOfStaffListItems; ++itemIdx)
```
Comment (JoshuaBonn1) [inline mscore/instrwidget.cpp]: Space around operator as per coding rules: `itemIdx = 0`

File: mscore/instrwidget.cpp [hunk 18]
```diff
@@ -478,6 +480,139 @@ void InstrumentsWidget::genPartList(Score* cs)
...
+                  
+                        int idx = parent->indexOfChild(item);
+                        int n = parent->childCount();
+                        // if staff item is not last of its part, move one slot down in part
+                        if ((idx + delta < n) && (idx + delta >= 0)) {
+                              StaffListItem* item1 = static_cast<StaffListItem*>(parent->takeChild(idx));
+                              // Qt looses the QComboBox set into StaffListItem when it is re-inserted into the tree:
+                              // get currently selected staff type and re-insert
+                              int staffTypeIdx = item1->staffTypeIdx();
+                              parent->insertChild(idx+delta, item1);
```
Comment (JoshuaBonn1) [inline mscore/instrwidget.cpp]: Space around operand

File: mscore/instrwidget.cpp [hunk 19]
```diff
@@ -590,85 +723,93 @@ void InstrumentsWidget::on_addButton_clicked()
...
-            if (parent->childCount() == 1) {
-                  Q_ASSERT(false); // shouldn't get here (remove button disabled when one item left)
-                  removeButton->setEnabled(false); // nevertheless, handle gracefully in release builds
-                  return;
-                  }
-            if (((StaffListItem*)item)->op() == ListItemOp::ADD) {
+      
+      for (QTreeWidgetItem* item : _partiturSelectedItems) {
+            
+
```
Comment (JoshuaBonn1) [inline mscore/instrwidget.cpp]: Extra space?

**Comment by allardnaber** [comment]
Finally seemed to manage it, sorry for my git stupidity. I'll review the other comments now and try to revert the revision.h (I was in doubt but saw it added in other PRs as well)
Thanks for your instructions!

**Comment by jthistle** [comment]
Good, the commit's correctly squashed. If you're going to make changes, commit them and squash again, just makes sure you use `HEAD~2` or `HEAD~n` when rebasing, where n is the number of commits you want to consider in the rebase.

**Comment by jthistle** [comment]
Also, it's a small thing, but might be quite tedious to fix: when you use the word `partitur`, what do you mean? I know it's German for 'score', but is that what it means in this context? I'm taking a tentative guess that you're German, also because you capitalized `Instruments` in the commit message :) If you're not, sorry, I don't want to cause any offense.

**Comment by allardnaber** [comment]
Haha not German, but close. I'm Dutch. Instruments was capitalized as it is the title of the dialog.
The already existing part list in the UI was called `partiturList` so I've used the same naming. 
I'm under the impression that this part wasn't updated for a while as there a lot of commented lines of code, as I wouldn't like to have non-English or unclear names (`witems`, `it`, `cs`).

I can have a look to improve this overaal, but I'll suggest to make it a separate PR, agree?

**Comment by jthistle** [comment]
Yes, sorry. If it already used the word partitur, then I guess it's fine to leave it as that. It's possible Werner wrote it and used a German variable name :)

File: mscore/instrwidget.cpp (hunk 17 above)
Comment (allardnaber) [reply mscore/instrwidget.cpp]: I have imported the project in QT Creator to resolve spacing and indentation issues. A lot of issues were already there. I missed a few in my own addition as XCode refuses to indent if it does not expect an indentation. Working on a few of the other comments now.

File: mscore/instrwidget.cpp (hunk 1 above)
Comment (allardnaber) [reply mscore/instrwidget.cpp]: It doesn't exist yet, just checked. Agree to put that in a separate PR later on?

File: mscore/instrwidget.h (hunk 14 above)
Comment (allardnaber) [reply mscore/instrwidget.h]: done

**Comment by allardnaber** [comment]
I've updated the PR, all comments were processed except for the elimination of static_cast.

File: mscore/instrwidget.cpp [hunk 20]
```diff
@@ -479,6 +481,137 @@ void InstrumentsWidget::genPartList(Score* cs)
...
+      // If children of multiple parents are selected, select all parents (and their children, by previous convention).
+      QSet<QTreeWidgetItem*> selectedParents;
+      bool containsParents = false;
+      for (QTreeWidgetItem* item : partiturList->selectedItems()) {
+            QTreeWidgetItem* parent = item->parent();
+            if (parent == NULL) {
+                  selectedParents.insert(item);
+                  containsParents = true;
+                  }
+            else {
```
Comment (JoshuaBonn1) [inline mscore/instrwidget.cpp]: Unneeded brackets

File: mscore/instrwidget.cpp [hunk 21]
```diff
@@ -479,6 +481,137 @@ void InstrumentsWidget::genPartList(Score* cs)
...
+                              }
+                        partiturList->setItemExpanded(item1, isExpanded);
+                        }
+
+                  }
+            }
+      else { // CHILDREN_SELECTED
+            QTreeWidgetItem* parent = itemsToMove.front()->parent();
+            int idx = delta > 0 ? parent->childCount() - 1 : 0;
+            for ( ; (item = parent->child(idx)); idx += -delta) {
```
Comment (JoshuaBonn1) [inline mscore/instrwidget.cpp]: Why not just use `idx -= delta`?

File: mscore/instrwidget.cpp [hunk 22]
```diff
@@ -479,6 +481,137 @@ void InstrumentsWidget::genPartList(Score* cs)
...
+                  
+                  int idx = parent->indexOfChild(item);
+                  int n = parent->childCount();
+                  // if staff item is not last of its part, move one slot down in part
+                  if ((idx + delta < n) && (idx + delta >= 0)) {
+                        StaffListItem* item1 = static_cast<StaffListItem*>(parent->takeChild(idx));
+                        // Qt looses the QComboBox set into StaffListItem when it is re-inserted into the tree:
+                        // get currently selected staff type and re-insert
+                        int staffTypeIdx = item1->staffTypeIdx();
+                        parent->insertChild(idx+delta, item1);
```
Comment (JoshuaBonn1) [inline mscore/instrwidget.cpp]: space around operand

File: mscore/instrwidget.cpp [hunk 23]
```diff
@@ -479,6 +481,137 @@ void InstrumentsWidget::genPartList(Score* cs)
...
+                        // after item has been inserted into the tree, create a new QComboBox and set its index
+                        item1->initStaffTypeCombo(true);
+                        item1->setStaffType(staffTypeIdx);
+                        }
+                  }
+            }
+      // restore the selected items
+      for (QTreeWidgetItem* item : originalSelection) {
```
Comment (JoshuaBonn1) [inline mscore/instrwidget.cpp]: Unneeded brackets

File: mscore/instrwidget.cpp [hunk 24]
```diff
@@ -590,85 +721,89 @@ void InstrumentsWidget::on_addButton_clicked()
...
                         }
                   else {
-                        parent->takeChild(parent->indexOfChild(item));
-                        delete item;
+                        ((StaffListItem*) item)->setOp(ListItemOp::I_DELETE);
+                        item->setHidden(true);
+                        
+                        // check if a staff is linked to this staff
+                        int idx = parent->indexOfChild(item);
+                        StaffListItem* sli = static_cast<StaffListItem*>(parent->child(idx+1));
```
Comment (JoshuaBonn1) [inline mscore/instrwidget.cpp]: space around operand

File: mscore/instrwidget.cpp [hunk 25]
```diff
@@ -590,85 +721,89 @@ void InstrumentsWidget::on_addButton_clicked()
...
+                        if (sli) {
+                              StaffListItem* sli2 = static_cast<StaffListItem*>(parent->child(idx+2));
```
Comment (JoshuaBonn1) [inline mscore/instrwidget.cpp]: space around operand

File: mscore/instrwidget.h [hunk 26]
```diff
@@ -31,14 +31,15 @@ class InstrumentGenre;
...
+enum class PartiturSelectionType : char { PARENTS_SELECTED, CHILDREN_SELECTED, NONE_SELECTED };
 
 //---------------------------------------------------------
 //   PartListItem
 //---------------------------------------------------------
 
 class PartListItem : public QTreeWidgetItem {
 
-   public:
+public:
```
Comment (JoshuaBonn1) [inline mscore/instrwidget.h]: All class scoping rules should be 3 spaces from the start of the class indentation (in this case 0 + 3 = 3 spaces). Make sure to undo this change and the following instances in this file.

File: mscore/instrwidget.cpp (hunk 20 above)
Comment (allardnaber) [reply mscore/instrwidget.cpp]: Isn't this _Exception 2_ under Braces on https://musescore.org/en/handbook/developers-handbook/finding-your-way-around/musescore-coding-rules?

Comment (JoshuaBonn1) [reply mscore/instrwidget.cpp]: Whoops! Missed the "if or else part". Ignore any further mention that falls under that category.

File: mscore/instrwidget.h (hunk 14 above)
Comment (dmitrio95) [reply mscore/instrwidget.h]: Um, no, I wouldn't confirm that :) I don't believe that this enumeration is used so frequently that storing it in one byte instead of 4 of 8 will save a lot of memory. So I would not bother making the underlying type as small as possible unless some measurements show that this really takes a lot of memory. Reverting that does not make much sense either though.

File: mscore/instrwidget.cpp (hunk 21 above)
Comment (allardnaber) [reply mscore/instrwidget.cpp]: You're right! will update

File: mscore/instrwidget.cpp (hunk 22 above)
Comment (allardnaber) [reply mscore/instrwidget.cpp]: these and the other 'space around operands' were already there. Fixed all that you commented on.

Comment (JoshuaBonn1) [reply mscore/instrwidget.cpp]: No reason not to improve what you touch, but okay.

File: mscore/instrwidget.cpp (hunk 1 above)
Comment (dmitrio95) [reply mscore/instrwidget.cpp]: Those macros apply only for `ScoreElement` descendants used in libmscore, and those assertions are possible at all just because of type identification system implemented in `ScoreElement` and its descendants.

Using `qobject_cast` here can make sense though since `StaffListItem` [is a QObject](https://github.com/musescore/MuseScore/blob/6226f66e5fd888245b06d210e9558ce49b5d0ff4/mscore/instrwidget.h#L58), and using `qobject_cast` allows run-time type checking for `QObject` descendants. But if it is known for sure that those tree items are `StaffListItem`s then this may be redundant or at least not needed to be changed urgently.

File: mscore/instrwidget.cpp (hunk 22 above)
Comment (allardnaber) [reply mscore/instrwidget.cpp]: You're right, I tried to be not to pedantic for my first contribution..

File: mscore/instrwidget.h (hunk 26 above)
Comment (allardnaber) [reply mscore/instrwidget.h]: Going to revert it. I applied to coding styles definition from the handbook in Qt Creator, but apparently this cannot be controlled correctly.

File: mscore/revision.h (hunk 15 above)
Comment (allardnaber) [reply mscore/revision.h]: removed

File: mscore/instrwidget.h (hunk 14 above)
Comment (jthistle) [reply mscore/instrwidget.h]: Oh, sorry. I thought just having a plain `enum` instead of `enum class` is bad practice, though? For example, see: https://stackoverflow.com/a/12581154
I agree that it makes little difference in terms of memory usage.

Comment (dmitrio95) [reply mscore/instrwidget.h]: `enum class` is indeed generally better than plain old `enum`, that is true. I didn't notice that there was a plain `enum` here and thought that the point of your comment was that `: char` part which seems to be useful in much more limited number of cases. But you are totally correct in pointing to `enum` vs `enum class` difference, sorry for missing that part!

Comment (allardnaber) [reply mscore/instrwidget.h]: It's now scoped. However, i'm still confused by the line above (`PART_LIST_ITEM, STAFF_LIST_ITEM`). Not sure of its purpose, it's used in constructing a `StaffListItem` (superclass of `QTreeWidgetItem`), but if I understand correctly it just passes `QTreeWidgetItem::UserType` in both cases. 
Removing the enum and using `QTreeWidgetItem::UserType` in the constructors does not cause any obvious problems. Any ideas?

File: mscore/instrwidget.cpp [hunk 27]
```diff
@@ -47,7 +47,7 @@ int StaffListItem::customTablatureIdx;
 //---------------------------------------------------------
 
 StaffListItem::StaffListItem(PartListItem* li)
-   : QTreeWidgetItem(li, STAFF_LIST_ITEM)
+      : QTreeWidgetItem(li, STAFF_LIST_ITEM)
```
Comment (dmitrio95) [inline mscore/instrwidget.cpp]: Continuing the indentation topic, changing these indents is also probably not needed, 3-spaces indents for initializer lists are commonly used in MuseScore, like in this case:
https://github.com/musescore/MuseScore/blob/eafe2e93683476e1fc719843201e8994d14c16d9/mscore/musescore.cpp#L976-L977

File: mscore/instrwidget.cpp [hunk 28]
```diff
@@ -88,20 +88,20 @@ void StaffListItem::initStaffTypeCombo(bool forceRecreate)
       // or in Instruments Wizard
       if (part) {
             const StringData* stringData = part->it ? &(part->it->stringData) :
-                        ( (part->part && part->part->instrument()) ? part->part->instrument()->stringData() : 0);
+                                                      ( (part->part && part->part->instrument()) ? part->part->instrument()->stringData() : 0);
```
Comment (dmitrio95) [inline mscore/instrwidget.cpp]: Not sure how large indents are usually used in this case but this change (and other similar changes) is anyway unrelated to this commit so it would be better not to include it here.

File: mscore/instrwidget.h (hunk 14 above)
Comment (allardnaber) [reply mscore/instrwidget.h]: Ah, I think it does. However, that differentiation is never used. I'll leave it untouched for now.

File: mscore/instrwidget.cpp (hunk 27 above)
Comment (allardnaber) [reply mscore/instrwidget.cpp]: How are these indents done? Are you manually indenting these? I just applied the code style schema as provided in the handbook, which triggered this change. It seems like it does not allow for 'half indents' in selected cases. same for the other indenting issue, it was caused by just applying the code style schema...

Comment (dmitrio95) [reply mscore/instrwidget.cpp]: Then this schema seems to not produce always correct results. Yes, I personally do these 3-space indents manually.

Comment (JoshuaBonn1) [reply mscore/instrwidget.cpp]: This probably should be added as a note next to the given schema in the handbook.

**Comment by allardnaber** [comment]
Are you still expecting me to do something before these branch can be merged?

**Comment by dmitrio95** [comment]
Well, at least indentation changes should probably be reverted since they are both irrelevant to this patch itself and not consistent with MuseScore coding style.

**Comment by dmitrio95** [comment]
Closing this for now, the coding style does not follow MuseScore guidelines. Please reopen if
